In [1]:
#import libraries
import pandas as pd
import numpy as np

In [2]:
#load dataset
path = r"C:\Users\Olasunkanmi Dara\Downloads\OSUN_crosschecked_GEO.csv"

df = pd.read_csv(path)

In [3]:
df.head(10)

State             LGA     Ward       PU-Code  \
0  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-001   
1  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-003   
2  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-004   
3  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-007   
4  OSUN  ATAKUMOSA EAST  IGANGAN  29-01-02-001   
5  OSUN  ATAKUMOSA EAST  IGANGAN  29-01-02-003   
6  OSUN  ATAKUMOSA EAST  IGANGAN  29-01-02-005   
7  OSUN  ATAKUMOSA EAST    IPOLE  29-01-03-001   
8  OSUN  ATAKUMOSA EAST    IPOLE  29-01-03-002   
9  OSUN  ATAKUMOSA EAST    IPOLE  29-01-03-003   

                          PU-Name  Accredited_Voters  Registered_Voters  \
0                 TOWN HALL IWARA                181                942   
1             L.A. SCHOOL, IWIKUN                245                881   
2  METHODIST PRY. SCHOOL, AYETORO                117                308   
3            OPEN SPACE ODO-IWARA                187                538   
4       N.U.D PRY. SCHOL, IGANGAN                187                558   
5  APOSTOLIC PRY. SCHOOL, OKO-AGO                204                724   
6              OPEN SPACE AJUMOBI                155                449   
7                TOWN HALL, IPOLE                292                787   
8             TOWN HALL ODO IJESA                225                909   
9            ST. ANDREW ISE-IJESA                243                565   

   Results_Found  Transcription_Count  Result_Sheet_Stamped  ...  \
0           True                   -1                 False  ...   
1           True                   -1                 False  ...   
2           True                   -1                 False  ...   
3           True                   -1                 False  ...   
4           True                   -1                 False  ...   
5           True                   -1                 False  ...   
6           True                   -1                 False  ...   
7           True                   -1                  True  ...   
8           True                   -1                 False  ...   
9           True                   -1                 False  ...   

   Result_Sheet_Invalid  Result_Sheet_Unclear  Result_Sheet_Unsigned  APC  \
0                 False                 False                UNKNOWN   44   
1                 False                 False                UNKNOWN   40   
2                 False                 False                UNKNOWN   14   
3                 False                 False                UNKNOWN   25   
4                 False                 False                UNKNOWN   22   
5                 False                 False                UNKNOWN    3   
6                 False                 False                UNKNOWN    4   
7                 False                 False                UNKNOWN  123   
8                 False                 False                UNKNOWN  113   
9                 False                 False                UNKNOWN   84   

    LP  PDP  NNPP                                       Results_File  \
0  273    0     0  https://inec-cvr-cache.s3.eu-west-1.amazonaws....   
1   11  175     1  https://inec-cvr-cache.s3.eu-west-1.amazonaws....   
2    3   99     1  https://inec-cvr-cache.s3.eu-west-1.amazonaws....   
3    4  155     0  https://docs.inecelectionresults.net/elections...   
4    0  161     0  https://docs.inecelectionresults.net/elections...   
5  199    0     0  https://docs.inecelectionresults.net/elections...   
6    1  148     0  https://docs.inecelectionresults.net/elections...   
7    1  152     0  https://docs.inecelectionresults.net/elections...   
8    5   98     0  https://docs.inecelectionresults.net/elections...   
9    3  129     0  https://docs.inecelectionresults.net/elections...   

    Latitude  Longitude  
0  7.4661824  4.6780746  
1  7.8463316  4.4879016  
2  7.5403179  4.8251988  
3  7.5467448  3.7881413  
4  7.4691347  4.5422036  
5  7.5063883  4.8264125  
6  7.5063883  4.8264125  
7  7.6239926  4.7331006  
8

In [4]:
df.shape

(2248, 21)

In [5]:
#quick view of the dataset
df.columns

Index(['State', 'LGA', 'Ward', 'PU-Code', 'PU-Name', 'Accredited_Voters',
       'Registered_Voters', 'Results_Found', 'Transcription_Count',
       'Result_Sheet_Stamped', 'Result_Sheet_Corrected',
       'Result_Sheet_Invalid', 'Result_Sheet_Unclear', 'Result_Sheet_Unsigned',
       'APC', 'LP', 'PDP', 'NNPP', 'Results_File', 'Latitude', 'Longitude'],
      dtype='object')

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2248 entries, 0 to 2247
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   State                   2248 non-null   object
 1   LGA                     2248 non-null   object
 2   Ward                    2248 non-null   object
 3   PU-Code                 2248 non-null   object
 4   PU-Name                 2248 non-null   object
 5   Accredited_Voters       2248 non-null   int64 
 6   Registered_Voters       2248 non-null   int64 
 7   Results_Found           2248 non-null   bool  
 8   Transcription_Count     2248 non-null   int64 
 9   Result_Sheet_Stamped    2248 non-null   bool  
 10  Result_Sheet_Corrected  2248 non-null   bool  
 11  Result_Sheet_Invalid    2248 non-null   bool  
 12  Result_Sheet_Unclear    2248 non-null   bool  
 13  Result_Sheet_Unsigned   2248 non-null   object
 14  APC                     2248 non-null   int64 
 15  LP  

In [7]:
#convert latitude and longitude to numeric

df['Latitude'] = pd.to_numeric(df['Latitude'], errors='coerce')
df['Longitude'] = pd.to_numeric(df['Longitude'], errors='coerce')


In [8]:
#check for nulls
df.isnull().sum()

State                      0
LGA                        0
Ward                       0
PU-Code                    0
PU-Name                    0
Accredited_Voters          0
Registered_Voters          0
Results_Found              0
Transcription_Count        0
Result_Sheet_Stamped       0
Result_Sheet_Corrected     0
Result_Sheet_Invalid       0
Result_Sheet_Unclear       0
Result_Sheet_Unsigned      0
APC                        0
LP                         0
PDP                        0
NNPP                       0
Results_File               0
Latitude                  37
Longitude                 37
dtype: int64

There are 2248 rows and 21 columns in this datasheet and we can see the Lat and Long columns have some missing values

Since the number of the missing rows are very small and insigniifcant, it is better to just drop them

In [9]:
df = df.dropna(subset=['Latitude', 'Longitude'])

In [10]:
df.isnull().sum()

State                     0
LGA                       0
Ward                      0
PU-Code                   0
PU-Name                   0
Accredited_Voters         0
Registered_Voters         0
Results_Found             0
Transcription_Count       0
Result_Sheet_Stamped      0
Result_Sheet_Corrected    0
Result_Sheet_Invalid      0
Result_Sheet_Unclear      0
Result_Sheet_Unsigned     0
APC                       0
LP                        0
PDP                       0
NNPP                      0
Results_File              0
Latitude                  0
Longitude                 0
dtype: int64

### Neighbour Identification

In [13]:
from geopy.distance import geodesic
from tqdm import tqdm  # corrected import statement

# Function to find neighboring polling units within a 1 km radius
def find_neighbours(df, radius=1):
    neighbours = []
    for index, row in tqdm(df.iterrows(), total=df.shape[0], desc='Processing units'):
        unit_location = (row['Latitude'], row['Longitude'])
        neighbour_units = df[df.apply(lambda x: geodesic(unit_location, (x['Latitude'], x['Longitude'])).km < radius, axis=1) & (df.index != index)]
        neighbours.append(neighbour_units.index.tolist())
        #neighbours.append(neighbour_units['PU-Code'].tolist())  # Replace 'PU-Code' with the actual column name for PU codes
    return neighbours

# Identify neighbors
df['Neighbours'] = find_neighbours(df)
df.head()


Processing units: 100%|██████████| 2211/2211 [17:26<00:00,  2.11it/s]


State             LGA     Ward       PU-Code  \
0  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-001   
1  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-003   
2  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-004   
3  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-007   
4  OSUN  ATAKUMOSA EAST  IGANGAN  29-01-02-001   

                          PU-Name  Accredited_Voters  Registered_Voters  \
0                 TOWN HALL IWARA                181                942   
1             L.A. SCHOOL, IWIKUN                245                881   
2  METHODIST PRY. SCHOOL, AYETORO                117                308   
3            OPEN SPACE ODO-IWARA                187                538   
4       N.U.D PRY. SCHOL, IGANGAN                187                558   

   Results_Found  Transcription_Count  Result_Sheet_Stamped  ...  \
0           True                   -1                 False  ...   
1           True                   -1                 False  ...   
2           True                   -1                 False  ...   
3           True                   -1                 False  ...   
4           True                   -1                 False  ...   

   Result_Sheet_Unclear  Result_Sheet_Unsigned  APC   LP  PDP  NNPP  \
0                 False                UNKNOWN   44  273    0     0   
1                 False                UNKNOWN   40   11  175     1   
2                 False                UNKNOWN   14    3   99     1   
3                 False                UNKNOWN   25    4  155     0   
4                 False                UNKNOWN   22    0  161     0   

                                        Results_File  Latitude Longitude  \
0  https://inec-cvr-cache.s3.eu-west-1.amazonaws....  7.466182  4.678075   
1  https://inec-cvr-cache.s3.eu-west-1.amazonaws....  7.846332  4.487902   
2  https://inec-cvr-cache.s3.eu-west-1.amazonaws....  7.540318  4.825199   
3  https://docs.inecelectionresults.net/elections...  7.546745  3.788141   
4  https://docs.inecelectionresults.net/elections...  7.469135  4.542204   

                                          Neighbours  
0                               [85, 88, 89, 90, 91]  
1                       [54, 61, 65, 644, 645, 1366]  
2                                                 []  
3                                                 []  
4  [47, 800, 822, 840, 865, 868, 943, 1000, 1461,...  

[5 rows x 22 columns]

In [14]:
df[['PU-Code', 'PU-Name', 'Latitude', 'Longitude', 'Neighbours']].head(10)

PU-Code                         PU-Name  Latitude  Longitude  \
0  29-01-01-001                 TOWN HALL IWARA  7.466182   4.678075   
1  29-01-01-003             L.A. SCHOOL, IWIKUN  7.846332   4.487902   
2  29-01-01-004  METHODIST PRY. SCHOOL, AYETORO  7.540318   4.825199   
3  29-01-01-007            OPEN SPACE ODO-IWARA  7.546745   3.788141   
4  29-01-02-001       N.U.D PRY. SCHOL, IGANGAN  7.469135   4.542204   
5  29-01-02-003  APOSTOLIC PRY. SCHOOL, OKO-AGO  7.506388   4.826412   
6  29-01-02-005              OPEN SPACE AJUMOBI  7.506388   4.826412   
7  29-01-03-001                TOWN HALL, IPOLE  7.623993   4.733101   
8  29-01-03-002             TOWN HALL ODO IJESA  0.000000   0.000000   
9  29-01-03-003            ST. ANDREW ISE-IJESA  7.622964   4.791227   

                                          Neighbours  
0                               [85, 88, 89, 90, 91]  
1                       [54, 61, 65, 644, 645, 1366]  
2                                                 []  
3                                                 []  
4  [47, 800, 822, 840, 865, 868, 943, 1000, 1461,...  
5          [6, 10, 12, 14, 16, 17, 23, 27, 39, 2184]  
6          [5, 10, 12, 14, 16, 17, 23, 27, 39, 2184]  
7  [252, 1172, 1179, 1182, 1205, 1219, 1220, 1221...  
8  [28, 29, 30, 31, 34, 35, 42, 43, 44, 66, 71, 7...  
9                                             [1180]

In [15]:
#rename Neighbours to Neighbouring units
df.rename(columns={'Neighbours': 'Neighbouring units'}, inplace=True)

# to check
df.head(10)

State             LGA     Ward       PU-Code  \
0  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-001   
1  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-003   
2  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-004   
3  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-007   
4  OSUN  ATAKUMOSA EAST  IGANGAN  29-01-02-001   
5  OSUN  ATAKUMOSA EAST  IGANGAN  29-01-02-003   
6  OSUN  ATAKUMOSA EAST  IGANGAN  29-01-02-005   
7  OSUN  ATAKUMOSA EAST    IPOLE  29-01-03-001   
8  OSUN  ATAKUMOSA EAST    IPOLE  29-01-03-002   
9  OSUN  ATAKUMOSA EAST    IPOLE  29-01-03-003   

                          PU-Name  Accredited_Voters  Registered_Voters  \
0                 TOWN HALL IWARA                181                942   
1             L.A. SCHOOL, IWIKUN                245                881   
2  METHODIST PRY. SCHOOL, AYETORO                117                308   
3            OPEN SPACE ODO-IWARA                187                538   
4       N.U.D PRY. SCHOL, IGANGAN                187                558   
5  APOSTOLIC PRY. SCHOOL, OKO-AGO                204                724   
6              OPEN SPACE AJUMOBI                155                449   
7                TOWN HALL, IPOLE                292                787   
8             TOWN HALL ODO IJESA                225                909   
9            ST. ANDREW ISE-IJESA                243                565   

   Results_Found  Transcription_Count  Result_Sheet_Stamped  ...  \
0           True                   -1                 False  ...   
1           True                   -1                 False  ...   
2           True                   -1                 False  ...   
3           True                   -1                 False  ...   
4           True                   -1                 False  ...   
5           True                   -1                 False  ...   
6           True                   -1                 False  ...   
7           True                   -1                  True  ...   
8           True                   -1                 False  ...   
9           True                   -1                 False  ...   

   Result_Sheet_Unclear  Result_Sheet_Unsigned  APC   LP  PDP  NNPP  \
0                 False                UNKNOWN   44  273    0     0   
1                 False                UNKNOWN   40   11  175     1   
2                 False                UNKNOWN   14    3   99     1   
3                 False                UNKNOWN   25    4  155     0   
4                 False                UNKNOWN   22    0  161     0   
5                 False                UNKNOWN    3  199    0     0   
6                 False                UNKNOWN    4    1  148     0   
7                 False                UNKNOWN  123    1  152     0   
8                 False                UNKNOWN  113    5   98     0   
9                 False                UNKNOWN   84    3  129     0   

                                        Results_File  Latitude Longitude  \
0  https://inec-cvr-cache.s3.eu-west-1.amazonaws....  7.466182  4.678075   
1  https://inec-cvr-cache.s3.eu-west-1.amazonaws....  7.846332  4.487902   
2  https://inec-cvr-cache.s3.eu-west-1.amazonaws....  7.540318  4.825199   
3  https://docs.inecelectionresults.net/elections...  7.546745  3.788141   
4  https://docs.inecelectionresults.net/elections...  7.469135  4.542204   
5  https://docs.inecelectionresults.net/elections...  7.506388  4.826412   
6  https://docs.inecelectionresults.net/elections...  7.506388  4.826412   
7  https://docs.inecelectionresults.net/elections...  7.623993  4.733101   
8  https://docs.inecelectionresults.net/elections...  0.000000  0.000000   
9  https://docs.inecelectionresults.net/elections...  7.622964  4.791227   

                                  Neighbouring units  
0                               [85, 88, 89, 90, 91]  
1                       [54, 61, 65, 644, 645, 1366]  
2                                                 []  
3                                    

### Outlier Score Calculation

In [16]:
def calculate_outlier_scores(df):
    parties = ['APC', 'LP', 'PDP', 'NNPP']  # Update this list with actual party columns in your dataset
    
    # Initialize columns for outlier scores
    for party in parties:
        df[f'{party}_Outlier_Score'] = 0
    
    for index, row in df.iterrows():
        neighbours = df.loc[row['Neighbouring units']]
        for party in parties:
            if not neighbours.empty:
                score = np.abs(row[party] - neighbours[party].mean())
                df.at[index, f'{party}_Outlier_Score'] = score
    
    return df

# Calculate outlier scores and update the DataFrame
df = calculate_outlier_scores(df)

# Display the first few rows of the updated DataFrame
df.head()


C:\Users\Olasunkanmi Dara\AppData\Local\Temp\ipykernel_13720\3293094367.py:13: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '34.8' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.at[index, f'{party}_Outlier_Score'] = score
C:\Users\Olasunkanmi Dara\AppData\Local\Temp\ipykernel_13720\3293094367.py:13: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '269.8' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.at[index, f'{party}_Outlier_Score'] = score
C:\Users\Olasunkanmi Dara\AppData\Local\Temp\ipykernel_13720\3293094367.py:13: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '77.4' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.at[index, f'{party}_Outlie

State             LGA     Ward       PU-Code  \
0  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-001   
1  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-003   
2  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-004   
3  OSUN  ATAKUMOSA EAST    IWARA  29-01-01-007   
4  OSUN  ATAKUMOSA EAST  IGANGAN  29-01-02-001   

                          PU-Name  Accredited_Voters  Registered_Voters  \
0                 TOWN HALL IWARA                181                942   
1             L.A. SCHOOL, IWIKUN                245                881   
2  METHODIST PRY. SCHOOL, AYETORO                117                308   
3            OPEN SPACE ODO-IWARA                187                538   
4       N.U.D PRY. SCHOL, IGANGAN                187                558   

   Results_Found  Transcription_Count  Result_Sheet_Stamped  ...  PDP  NNPP  \
0           True                   -1                 False  ...    0     0   
1           True                   -1                 False  ...  175     1   
2           True                   -1                 False  ...   99     1   
3           True                   -1                 False  ...  155     0   
4           True                   -1                 False  ...  161     0   

                                        Results_File  Latitude  Longitude  \
0  https://inec-cvr-cache.s3.eu-west-1.amazonaws....  7.466182   4.678075   
1  https://inec-cvr-cache.s3.eu-west-1.amazonaws....  7.846332   4.487902   
2  https://inec-cvr-cache.s3.eu-west-1.amazonaws....  7.540318   4.825199   
3  https://docs.inecelectionresults.net/elections...  7.546745   3.788141   
4  https://docs.inecelectionresults.net/elections...  7.469135   4.542204   

                                  Neighbouring units  APC_Outlier_Score  \
0                               [85, 88, 89, 90, 91]          34.800000   
1                       [54, 61, 65, 644, 645, 1366]          40.333333   
2                                                 []           0.000000   
3                                                 []           0.000000   
4  [47, 800, 822, 840, 865, 868, 943, 1000, 1461,...          88.700000   

   LP_Outlier_Score PDP_Outlier_Score  NNPP_Outlier_Score  
0             269.8         77.400000            0.200000  
1               7.5         71.166667            1.666667  
2               0.0          0.000000            0.000000  
3               0.0          0.000000            0.000000  
4              12.9         61.000000            0.500000  

[5 rows x 26 columns]

### Sort and Report

In [17]:
sorted_apc = df.sort_values(by='APC_Outlier_Score', ascending=False).head(3)
sorted_lp = df.sort_values(by='LP_Outlier_Score', ascending=False).head(3)
sorted_pdp = df.sort_values(by='PDP_Outlier_Score', ascending=False).head(3)
sorted_nnpp = df.sort_values(by='NNPP_Outlier_Score', ascending=False).head(3)


# Display the top 3 outliers for each party
print("Top 3 APC Outliers:")
print(sorted_apc)
print("\nTop 3 LP Outliers:")
print(sorted_lp)
print("\nTop 3 PDP Outliers:")
print(sorted_pdp)
print("\nTop 3 NNPP Outliers:")
print(sorted_nnpp)


Top 3 APC Outliers:
    State        LGA                 Ward       PU-Code  \
290  OSUN     BORIPE  COLLEGE/EGBADA ROAD  29-06-03-009   
364  OSUN  EDE NORTH      OLOGUN/AGBAAKIN  29-07-03-004   
415  OSUN  EDE NORTH    ISIBO/BUARI-ISOLA  29-07-08-009   

                                               PU-Name  Accredited_Voters  \
290                               L.A. SCHOOL, EESADEE                394   
364  OPEN SPACE IN FRONT OF BAALE ILE ARANYIN( ARAN...                332   
415                        56, AKODA COMP.S AGATE ROAD                159   

     Registered_Voters  Results_Found  Transcription_Count  \
290                951           True                   -1   
364                750           True                   -1   
415                343           True                   -1   

     Result_Sheet_Stamped  ...  PDP  NNPP  \
290                 False  ...   67     0   
364                 False  ...   23     0   
415                 False  ...  124     0   

    

In [18]:

# Save the DataFrame to CSV with specified data types
df.to_csv('Final_GEO.csv', index=False)

### Excel file with the sorted outliers scores

In [26]:

# Function to create a DataFrame for each party with the specified columns
def create_outlier_df(party, df):
    columns = ['PU-Code', 'PU-Name', 'Party', 'Outlier_Score', 'Neighbouring units']
    outlier_df = df[['PU-Code', 'PU-Name', f'{party}_Outlier_Score', 'Neighbouring units']].copy()
    outlier_df = outlier_df[outlier_df[f'{party}_Outlier_Score'] > 0]  # Only include rows with non-zero outlier scores
    outlier_df['Party'] = party
    outlier_df = outlier_df.rename(columns={f'{party}_Outlier_Score': 'Outlier_Score'})
    return outlier_df[columns]

# Create DataFrames for each party
apc_outliers = create_outlier_df('APC', df)
lp_outliers = create_outlier_df('LP', df)
pdp_outliers = create_outlier_df('PDP', df)
nnpp_outliers = create_outlier_df('NNPP', df)

# Sort each DataFrame by 'Outlier_Score'
sorted_apc = apc_outliers.sort_values(by='Outlier_Score', ascending=False)
sorted_lp = lp_outliers.sort_values(by='Outlier_Score', ascending=False)
sorted_pdp = pdp_outliers.sort_values(by='Outlier_Score', ascending=False)
sorted_nnpp = nnpp_outliers.sort_values(by='Outlier_Score', ascending=False)

# Save each sorted DataFrame to an Excel file
with pd.ExcelWriter(r'C:\Users\Olasunkanmi Dara\Downloads\sorted_outlier_scores.xlsx') as writer:
    sorted_apc.to_excel(writer, sheet_name='APC_Outliers', index=False)
    sorted_lp.to_excel(writer, sheet_name='LP_Outliers', index=False)
    sorted_pdp.to_excel(writer, sheet_name='PDP_Outliers', index=False)
    sorted_nnpp.to_excel(writer, sheet_name='NNPP_Outliers', index=False)



